sonar/resources/projects/api.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Swiss Open Access Repository
+# Copyright (C) 2021 RERO
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""API for projects resources."""
+
+from invenio_pidstore.providers.recordid import RecordIdProvider
+from invenio_records.dumpers import ElasticsearchDumper, ElasticsearchDumperExt
+from invenio_records.systemfields import ConstantField
+from invenio_records_resources.records.api import Record
+from invenio_records_resources.records.systemfields import IndexField, PIDField
+from invenio_records_resources.services.records.components import \
+    ServiceComponent
+
+from sonar.affiliations import AffiliationResolver
+from sonar.modules.organisations.api import OrganisationRecord
+from sonar.modules.users.api import UserRecord
+
+from . import models
+
+
+class ElasticsearchDumperObjectsExt(ElasticsearchDumperExt):
+    """Interface for Elasticsearch dumper extensions."""
+
+    def dump(self, record, data):
+        """Dump the data for indexing."""
+        if data['metadata'].get('user'):
+            data['metadata']['user'] = {
+                'pid':
+                UserRecord.get_pid_by_ref_link(
+                    data['metadata']['user']['$ref'])
+            }
+
+        if data['metadata'].get('organisation'):
+            data['metadata']['organisation'] = {
+                'pid':
+                OrganisationRecord.get_pid_by_ref_link(
+                    data['metadata']['organisation']['$ref'])
+            }
+
+
+class ProjectRecord(Record):
+    """API for projects resources."""
+
+    # Configuration
+    model_cls = models.ProjectRecordMetadata
+
+    # System fields
+    schema = ConstantField(
+        '$schema', 'https://sonar.ch/schemas/projects/project-v1.0.0.json')
+
+    index = IndexField('projects-project-v1.0.0', search_alias='projects')
+
+    pid = PIDField('id', pid_type='proj', provider=RecordIdProvider)
```

Comment:
In other project, we need a specific provider to avoid pid generation shared between resources: i.e. `/api/documents/1` and `/api/projects/1`.

Reply:
OK, I will create another issue to update that, if it's necessary.